packages/material-ui/src/ClickAwayListener/ClickAwayListener.js

```diff
@@ -76,21 +82,30 @@ function ClickAwayListener(props) {
     movedRef.current = true;
   }, []);
 
-  const listenerProps = {};
-  if (mouseEvent !== false) {
-    listenerProps[mouseEvent] = handleClickAway;
-  }
-  if (touchEvent !== false) {
-    listenerProps[touchEvent] = handleClickAway;
-    listenerProps.onTouchMove = handleTouchMove;
-  }
+  React.useEffect(() => {
+    const mappedMouseEvent = mapEventPropToEvent(mouseEvent);
+    const mappedTouchEvent = mapEventPropToEvent(touchEvent);
 
-  return (
-    <React.Fragment>
-      {React.cloneElement(children, { ref: handleRef })}
-      <EventListener target="document" {...listenerProps} />
-    </React.Fragment>
-  );
+    if (mouseEvent !== false) {
+      document.addEventListener(mappedMouseEvent, handleClickAway);
+    }
+    if (touchEvent !== false) {
+      document.addEventListener(mappedTouchEvent, handleClickAway);
+      document.addEventListener('touchmove', handleTouchMove);
+    }
+
+    return () => {
+      if (mouseEvent !== false) {
+        document.removeEventListener(mouseEvent, handleClickAway);
+      }
+      if (touchEvent !== false) {
+        document.removeEventListener(touchEvent, handleClickAway);
+        document.removeEventListener('touchmove', handleTouchMove);
+      }
+    };
+  }, [handleClickAway, handleTouchMove, mouseEvent, touchEvent]);
```

Comment:
I think we can split those into multiple effects which means we don't have to conditionally remove the eventlisteners but conditionally return a cleanup function.